Add an Azure CloudProvider Implementation

colemickens opened the pull request:
This PR adds `Azure` as a cloudprovider provider for Kubernetes. It specifically adds support for native pod networking (via Azure User Defined Routes) and L4 Load Balancing (via Azure Load Balancers).

I did have to add `clusterName` as a parameter to the `LoadBalancers` methods. This is because Azure only allows one "LoadBalancer" object per set of backend machines. This means a single "LoadBalancer" object must be shared across the cluster. The "LoadBalancer" is named via the `cluster-name` parameter passed to `kube-controller-manager` so as to enable multiple clusters per resource group if the user desires such a configuration.

There are few things that I'm a bit unsure about:
1. The implementation of the `Instances` interface. It's not extensively documented, it's not really clear what the different functions are used for, and my questions on the ML didn't get an answer.
2. Counter to the comments on the `LoadBalancers` Interface, I modify the `api.Service` object in `EnsureLoadBalancerDeleted`, but not with the intention of affecting Kube's view of the Service. I simply do it so that I can remove the `Port`s on the `Service` object and then re-use my reconciliation logic that can handle removing stale/deleted Ports. 
3. The logging is a bit verbose. I'm looking for guidance on the appropriate log level to use for the chattier bits.

Due to the (current) lack of Instance Metadata Service and lack of Virtual Machine Identity in Azure, the user is required to do a few things to opt-in to this provider. These things are called-out as they are in contrast to AWS/GCE:
1. The user must provision an Azure Active Directory ServicePrincipal with `Contributor` level access to the resource group that the cluster is deployed in. This creation process is documented [by Hashicorp](https://www.packer.io/docs/builders/azure-setup.html) or [on the MSDN Blog](https://blogs.msdn.microsoft.com/arsen/2016/05/11/how-to-create-and-test-azure-service-principal-using-azure-cli/).
2. The user must place a JSON file somewhere on each Node that conforms to the `AzureConfig` struct defined in `azure.go`. (This is automatically done in the Azure flavor of [Kubernetes-Anywhere](https://github.com/kubernetes/kubernetes-anywhere).)
3. The user must specify `--cloud-config=/path/to/azure.json` as an option to `kube-apiserver` and `kube-controller-manager` similarly to how the user would need to pass `--cloud-provider=azure`.

I've been running approximately this code for a month and a half. I only encountered one bug which has since been fixed and covered by a unit test. I've just deployed a new cluster (and a Type=LoadBalancer nginx Service) using this code (via `kubernetes-anywhere`) and have posted [the `kube-controller-manager` logs](https://gist.github.com/colemickens/1bf6a26e7ef9484a72a30b1fcf9fc3cb) for anyone who is interested in seeing the logs of the logic.

If you're interested in this PR, you can use the instructions in my [`azure-kubernetes-demo` repository](https://github.com/colemickens/azure-kubernetes-demo) to deploy a cluster with minimal effort via [`kubernetes-anywhere`](https://github.com/kubernetes/kubernetes-anywhere). (There is currently [a pending PR in `kubernetes-anywhere` that is needed](https://github.com/kubernetes/kubernetes-anywhere/pull/172) in conjuncture with this PR). I also have a pre-built `hyperkube` image: `docker.io/colemickens/hyperkube-amd64:v1.4.0-alpha.0-azure`, which will be kept in sync with the branch this PR stems from.

I'm hoping this can land in the Kubernetes 1.4 timeframe.

CC (potential code reviewers from Azure): @ahmetalpbalkan @brendandixon @paulmey

CC (other interested Azure folk): @brendandburns @johngossman @anandramakrishna @jmspring @jimzim

CC (others who've expressed interest): @codefx9 @edevil @thockin @rootfs

colemickens committed:
add clusterName to Loadbalancer methods

colemickens committed:
implement azure cloudprovider

colemickens committed:
godep: add azure-sdk-for-go, go-autorest

colemickens committed:
vendor azure-sdk-for-go, go-autorest

brendandburns commented on pkg/cloudprovider/cloud.go:
```diff
@@ -81,21 +81,21 @@ type LoadBalancer interface {
 	// GetLoadBalancer returns whether the specified load balancer exists, and
 	// if so, what its status is.
 	// Implementations must treat the *api.Service parameter as read-only and not modify it.
```
Document the new parameter please.

colemickens replied:
I did document it on each method on the Interface, github just didn't notice properly here.

brendandburns commented on pkg/cloudprovider/providers/azure/azure.go:
```diff
@@ -0,0 +1,143 @@
+const AzureProviderName = "azure"
+
+type AzureConfig struct {
+	Cloud             string `json:"cloud"`
```
yaml tags too?

colemickens replied:
And presumably change the json decode in `azure.go` to yaml so we can allow both; sounds good.

colemickens replied:
Actually, it looks like both https://github.com/ghodss/yaml and http://godoc.org/gopkg.in/yaml.v2 are in Godeps, but ghodss flavor is used more (not a scientific study). It also seems to translate yaml->json and then decodes into the struct, obviating the need for yaml tags, but I suppose it can't hurt and I'll add tests either way.

brendandburns commented on pkg/cloudprovider/providers/azure/azure.go:
```diff
@@ -0,0 +1,143 @@
+}
+
+func init() {
+	cloudprovider.RegisterCloudProvider(AzureProviderName, func(configReader io.Reader) (cloudprovider.Interface, error) {
```
move the anonymous function out to be a defined function (and add tests ;)

brendandburns commented on pkg/cloudprovider/providers/azure/azure.go:
```diff
@@ -0,0 +1,143 @@
+
+		oauthConfig, err := az.Environment.OAuthConfigForTenant(az.TenantID)
+		if err != nil {
+			glog.Errorf("init: failed to determine oauth configuration")
```
In general, if I'm returning an error, I generally don't log, I figure that's the caller's responsibility.

brendandburns commented on pkg/cloudprovider/providers/azure/azure_instances.go:
```diff
@@ -0,0 +1,90 @@
+
+// List lists instances that match 'filter' which is a regular expression which must match the entire instance name (fqdn)
+func (az *AzureCloud) List(filter string) ([]string, error) {
+	return nil, fmt.Errorf("not supported")
```
why not implement this in this function using a list all method?

colemickens replied:
Hm, do I attempt to distinguish Kubernetes Node VMs from other "regular" VMs in the resource group? Is there a convention for tagging VMs and then filtering on the tag and then filtering on the `filter` regex string?

colemickens replied:
Maybe something like if a tag exists: `"k8s.io/clusterName": "{clusterName}"` then include it as part of the list to filter with `filter`?

brendandburns commented on pkg/cloudprovider/providers/azure/azure_loadbalancer.go:
```diff
@@ -0,0 +1,623 @@
+	lbName := getLoadBalancerName(clusterName)
+	pipName := getPublicIPName(clusterName, service)
+	serviceName := getServiceName(service)
+	glog.Infof("get(%s): START clusterName=%q lbName=%q", serviceName, clusterName, lbName)
```
These should be at least glog.V(2) if not glog.V(> 2)

brendandburns commented on pkg/cloudprovider/providers/azure/azure_loadbalancer.go:
```diff
@@ -0,0 +1,623 @@
+	glog.Infof("get(%s): START clusterName=%q lbName=%q", serviceName, clusterName, lbName)
+
+	glog.Infof("get(%s): lb(%s) - retrieving", serviceName, lbName)
+	_, err = az.LoadBalancerClient.Get(az.ResourceGroup, lbName, "")
```
looks like you aren't checking the `err` and you are dropping the other result, is this call needed?

brendandburns replied:
oh, now I see you are passing it to the `check` function.  Please use different names for clarity.  (here and elsewhere)

brendandburns commented on pkg/cloudprovider/providers/azure/azure_loadbalancer.go:
```diff
@@ -0,0 +1,623 @@
+	glog.Infof("get(%s): lb(%s) - retrieving", serviceName, lbName)
+	_, err = az.LoadBalancerClient.Get(az.ResourceGroup, lbName, "")
+	if existsLb, err := checkResourceExistsFromError(err); err != nil {
+		glog.Errorf("get(%s): lb(%s) - retrieving failed: %q", serviceName, lbName, err)
```
as elsewhere wrt to both logging and returning an error.

brendandburns commented on pkg/cloudprovider/providers/azure/azure_loadbalancer.go:
```diff
@@ -0,0 +1,623 @@
+		glog.Errorf("get(%s): lb(%s) - retrieving failed: %q", serviceName, lbName, err)
+		return nil, false, err
+	} else if !existsLb {
+		glog.Infof("get(%s): lb(%s) - doesn't exist", serviceName, lbName)
```
definitely glog.v(2) at least.

brendandburns commented on pkg/cloudprovider/providers/azure/azure_loadbalancer.go:
```diff
@@ -0,0 +1,623 @@
+		return nil, false, nil
+	}
+
+	glog.Infof("get(%s): pip(%s) - retrieving", serviceName, pipName)
```
glog.vp(5)

brendandburns commented on pkg/cloudprovider/providers/azure/azure_loadbalancer.go:
```diff
@@ -0,0 +1,623 @@
+		return nil, false, nil
+	}
+
+	glog.Infof("get(%s): FINISH")
```
delete.

brendandburns commented on pkg/cloudprovider/providers/azure/azure_loadbalancer.go:
```diff
@@ -0,0 +1,623 @@
+		_, err := az.SecurityGroupsClient.CreateOrUpdate(az.ResourceGroup, *sg.Name, sg, nil)
+		if err != nil {
+			glog.Errorf("ensure(%s): sg(%s) - updating failed: %q", serviceName, *sg.Name, err)
+			return nil, fmt.Errorf("failed to update security group. err=%q", err)
```
don't wrap the error, just return it.

brendandburns commented on pkg/cloudprovider/providers/azure/azure_loadbalancer.go:
```diff
@@ -0,0 +1,623 @@
+			if strings.EqualFold(*config.Name, lbFrontendIPConfigName) {
+				glog.Infof("reconcile(%s)(%t): lb frontendconfig(%s) - dropping", serviceName, wantLb, lbFrontendIPConfigName)
+				newConfigs = append(newConfigs[:i],
+					newConfigs[i+1:]...)
```
nit: remove line break.

brendandburns commented on pkg/cloudprovider/providers/azure/azure_loadbalancer.go:
```diff
@@ -0,0 +1,623 @@
+		return fmt.Errorf("failed to retrieve vm instance. instance=%q", machineName)
+	}
+
+	primaryNicID, err := getPrimaryNicID(machine)
```
by general golang style this should be `getPrimaryNICID` I realize `NICID` might be confusing, so perhaps `getPrimaryIDForNIC`?

colemickens replied:
`getPrimaryNICID` is accurate though, while I feel `getPrimaryIDForNIC` is not. (It's retrieving the NIC ID, and passing in the machine object, rather than getting the NIC object by it's ID as I feel `getPrimaryIDForNIC` would imply).

Maybe something like `getPrimaryNetworkInterfaceID` since the `azure-sdk-for-go` already refers to the NICs as "Interface".

ahmetalpbalkan commented on pkg/cloudprovider/providers/azure/azure.go:
```diff
@@ -0,0 +1,143 @@
+		switch az.Cloud {
+		case "fairfax":
+			az.Environment = azure.USGovernmentCloud
+		case "mooncake":
```
these are most definitely internal-only code names for these regions.

colemickens replied:
Agreed, let's sort it out here and then I'll consume it: https://github.com/Azure/go-autorest/pull/80

ahmetalpbalkan commented on pkg/cloudprovider/providers/azure/azure_instances.go:
```diff
@@ -0,0 +1,90 @@
+// Note that if the instance does not exist or is no longer running, we must return ("", cloudprovider.InstanceNotFound)
+func (az *AzureCloud) InstanceID(name string) (string, error) {
+	machine, err := az.VirtualMachinesClient.Get(az.ResourceGroup, name, "")
+	if existsMachine, err := checkResourceExistsFromError(err); err != nil {
```
lots of `az.VirtualMachinesClient.Get` then `checkResourceExistsFromError`, maybe extract this into a utility function?

ahmetalpbalkan commented on pkg/cloudprovider/providers/azure/azure_loadbalancer.go:
```diff
@@ -0,0 +1,623 @@
+	lbBackendPoolName := getBackendPoolName(clusterName)
+	lbBackendPoolID := az.getBackendPoolID(lbName, lbBackendPoolName)
+
+	wantLb := (len(service.Spec.Ports) > 0)
```
parens not necessary

brendandburns commented on pkg/cloudprovider/providers/azure/azure_routes.go:
```diff
@@ -0,0 +1,189 @@
+	}
+	nicName := getLastSegment(nicID)
+
+	nic, err := az.InterfacesClient.Get(
```
nit:  this formatting is weird, doesn't it all fit on one line?

brendandburns commented on pkg/cloudprovider/providers/azure/azure_util.go:
```diff
@@ -0,0 +1,206 @@
+// returns the full identifier of a machine
+func (az *AzureCloud) getMachineID(machineName string) string {
+	return fmt.Sprintf(
+		"/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Compute/virtualMachines/%s",
```
nit: maybe extract these out into consts?

brendandburns commented on pkg/cloudprovider/providers/azure/azure_util.go:
```diff
@@ -0,0 +1,206 @@
+
+// returns the deepest child's identifier from a full identifier string.
+func getLastSegment(ID string) string {
+	parts := strings.Split(ID, "/")
```
should you check if len(parts) is zero here?

brendandburns commented on pkg/cloudprovider/providers/azure/azure_util.go:
```diff
@@ -0,0 +1,206 @@
+
+// returns the equivalent LoadBalancerRule, SecurityRule and LoadBalancerProbe
+// protocol types for the given Kubernetes protocol type.
+func getProtosFromKubeProto(protocol api.Protocol) (network.TransportProtocol, network.SecurityRuleProtocol, network.ProbeProtocol, error) {
```
nit: `getProtocolFromKubernetesProtocol`

brendandburns commented on pkg/cloudprovider/providers/azure/azure_util.go:
```diff
@@ -0,0 +1,206 @@
+// returns the equivalent LoadBalancerRule, SecurityRule and LoadBalancerProbe
+// protocol types for the given Kubernetes protocol type.
+func getProtosFromKubeProto(protocol api.Protocol) (network.TransportProtocol, network.SecurityRuleProtocol, network.ProbeProtocol, error) {
+	if protocol == api.ProtocolTCP {
```
total nit, but I feel like:
```
switch protocol {
  case api.ProtocolTCP:
    ...
  default:
    ...
}
```

is more self-documenting.

brendandburns commented on pkg/cloudprovider/providers/azure/azure_util.go:
```diff
@@ -0,0 +1,206 @@
+func getPrimaryNicID(machine compute.VirtualMachine) (string, error) {
+	var nicRef *compute.NetworkInterfaceReference
+
+	if len(*machine.Properties.NetworkProfile.NetworkInterfaces) == 1 {
```
personally, I would structure this:
```
if (...) {
   return &(*machine...)
} else {
   for {
     if *ref.Properties.Primary {
         return &ref
    }
 }
return nil, err
}

Your call, though.

brendandburns commented on pkg/cloudprovider/providers/azure/azure_util.go:
```diff
@@ -0,0 +1,206 @@
+		return nil, fmt.Errorf("cannot determine primary ipconfig")
+	}
+
+	if ipconfigRef == nil {
```
as above wrt early return and removing this if conditional.

brendandburns commented on pkg/cloudprovider/providers/azure/azure_util.go:
```diff
@@ -0,0 +1,206 @@
+
+func getRuleName(service *api.Service, port api.ServicePort) string {
+	return fmt.Sprintf("%s-%s-%d-%d",
+		getRulePrefix(service),
```
indent more to match the first parameter?

ahmetalpbalkan commented on pkg/cloudprovider/providers/azure/azure.go:
```diff
@@ -0,0 +1,143 @@
+	SecurityGroupName string `json:"securityGroupName"`
+	RouteTableName    string `json:"routeTableName"`
+
+	AdClientID     string `json:"adClientId"`
```
Ad --> AAD

brendandburns commented on pkg/cloudprovider/providers/azure/azure_zones.go:
```diff
@@ -0,0 +1,54 @@
+}
+
+func (az *AzureCloud) GetZone() (cloudprovider.Zone, error) {
+	if faultDomain == nil {
```
do you need to protect this with a lock?  Is there a possible race here?

brendandburns commented:
Generally this looks ok to me.

Three major over-arching comments.

1) I don't think that any of the `glog.Infof` should be vanilla `Infof` they're way too spammy.  Use `glog.V(...)` and in some cases I think the comments should simply be removed (or set to `glog.V(10)`).

2) I agree with Ahmet that the `err = ...` and `checkErrNotFound` idiom is a little disatisfying.  I'd see if it can be cleaned up.

3) you need to run gofmt, govet and fix up the boilerplate in a bunch of files.  Look at the Jenkins verification test output for more details.  Likewise, it looks like you need to update the `FakeCloud` object used in testing to know about cluster name.

Thanks!
--brendan

ahmetalpbalkan commented on pkg/cloudprovider/providers/azure/azure.go:
```diff
@@ -0,0 +1,143 @@
+			az.AdClientSecret,
+			az.Environment.ServiceManagementEndpoint)
+		if err != nil {
+			glog.Errorf("azurecp:init: failed to create service principal token")
```
same as Brendan's comment above, just return the err (preferably wrap?)

colemickens replied:
Preferred way to wrap? Just fmt.Errorf("some text: %q", err) ?

colemickens replied:
Decided to err on the side of less wrapping...

ahmetalpbalkan commented on pkg/cloudprovider/providers/azure/azure_routes.go:
```diff
@@ -0,0 +1,189 @@
+			}
+		}
+	} else {
+		kubeRoutes = []*cloudprovider.Route{}
```
why doesn't returning a nil work here? (i.e. why do we need this else block)

ahmetalpbalkan commented on pkg/cloudprovider/providers/azure/azure_routes.go:
```diff
@@ -0,0 +1,189 @@
+			Properties: &network.RouteTablePropertiesFormat{},
+		}
+
+		_, err = az.RouteTablesClient.CreateOrUpdate(az.ResourceGroup, az.RouteTableName, routeTable, nil)
```
shorthand form? `if err := .. ; err!= nil {return err}` here and below?

colemickens replied:
probably going to punt on this, shorthand form doesn't save any lines unless I have an even longer (too long, imo) line here.

ahmetalpbalkan commented on pkg/cloudprovider/providers/azure/azure_routes.go:
```diff
@@ -0,0 +1,189 @@
+	if existsSubnet, err := checkResourceExistsFromError(err); err != nil {
+		return err
+	} else if !existsSubnet {
+		glog.Infof("create: subnet was unexpectedly nil! vnet=%q subnet=%q", az.VnetName, az.SubnetName)
```
info or error/warning?

colemickens replied:
Yeah, just realized glog has `Warning`. I assumed there was no `Warn` level at all since Warnf/Warnln are absent. I'll change to `Warning()`.

colemickens commented:
Thanks @brendandburns, @ahmetalpbalkan for reviewing.

To the over-arching comments:

1. Removed all vanilla `glog.Info` and followed these guidelines:
   - `V(2)` for entry/exit of non-spammy functions (EnsureLoadBalancer{Created,Deleted}, {Create,Delete}Route)
   - `V(3)` for creation/deletion of Azure resources
   - `V(5)` for GetLoadBalancer exists=false exits
   - `V(10)` for internal logic (reconciliation) and ListRoutes
   - removed unnecessary Infos for GETs

2. Add helpers to combine the err/exists check (`checkErrNotFound` idiom)

3. Ran the `gofmt`/`govet`/`golint` and a normal full release.

Outstanding:

1. How to implement `List` for `Instances`? Add `clusterName` parameter to filter out instances on if they're tagged `"kubernetes.io/clusterName": "{clusterName}"`? Or don't try to filter non-Kubernetes nodes in the resource group?

2. Want to wait on this PR: https://github.com/Azure/go-autorest/pull/80

3. Squash after reviewers are happy.

Misc:

I'm now rebased on master (as of the last hour or so) and am now pushing hyperkube images tagged as: `docker.io/colemickens/hyperkube-amd64:v1.4.0-alpha.1-azure`.

rootfs commented on pkg/cloudprovider/providers/azure/azure.go:
```diff
@@ -0,0 +1,196 @@
+/*
+Copyright 2014 The Kubernetes Authors.
```
2016

mikedanese commented on pkg/cloudprovider/cloud.go:
```diff
@@ -81,21 +81,25 @@ type LoadBalancer interface {
 	// Implementations must treat the *api.Service parameter as read-only and not modify it.
-	EnsureLoadBalancerDeleted(service *api.Service) error
+	// Parameter 'clusterName' is the name of the cluster as presented to kube-controller-manager
+	EnsureLoadBalancerDeleted(clusterName string, service *api.Service) error
```
Yet more change to the cloudprovider interface required to accommodate new clouds which indicates that it was poorly designed. cc @davidopp @bgrant0607

colemickens replied:
Somewhat related, my comment here:  https://github.com/kubernetes/kubernetes/issues/24343#issuecomment-233022889

mikedanese commented:
@colemickens please add an OWNERS file to pkg/cloudprovider/providers/azure

colemickens commented:
@mikedanese added.
@brendandburns Do you want to be listed in OWNERS as well?

brendandburns commented:
Sure, add me to OWNERS too, I'll try to do another review pass tomorrow morning.

brendandburns commented on pkg/cloudprovider/providers/azure/azure_loadbalancer.go:
```diff
@@ -0,0 +1,610 @@
+		}
+	}
+
+	lbNeedsCreate := false
```
nit: this variable isn't really necessary, you can just use `!existsLb`

colemickens commented on pkg/cloudprovider/providers/azure/azure_loadbalancer.go:
```diff
@@ -0,0 +1,610 @@
+			Properties: &network.LoadBalancerPropertiesFormat{},
+		}
+		lbNeedsCreate = true
+		glog.V(3).Infof("ensure(%s): lb(%s) - needs creation")
```
(will drop)

brendandburns commented on pkg/cloudprovider/providers/azure/azure_loadbalancer.go:
```diff
@@ -0,0 +1,610 @@
+	lbBackendPoolID := az.getBackendPoolID(lbName, lbBackendName)
+	hostUpdates := make([]func() error, len(hosts))
+	for i, host := range hosts {
+		f := func(serviceName, host, lbBackendPoolID string) func() error {
```
This construction is needlessly complicated.  serviceName and lbBackendPoolID never change so you don't need to pass them around.  and if you set:
`localHost := host`

you won't need the `func(...)` indirection at all.

colemickens commented on pkg/cloudprovider/providers/azure/azure_loadbalancer.go:
```diff
@@ -0,0 +1,610 @@
+
+	sg, err := az.SecurityGroupsClient.Get(az.ResourceGroup, az.SecurityGroupName, "")
+	if err != nil {
+		glog.Errorf("ensure(%s): sg(%s) - retrieving failed: %q", serviceName, *sg.Name, err)
```
(will drop)

brendandburns commented on pkg/cloudprovider/providers/azure/azure_loadbalancer.go:
```diff
@@ -0,0 +1,610 @@
+		if serviceOwnsRule(service, *existingProbe.Name) {
+			glog.V(10).Infof("reconcile(%s)(%t): lb probe(%s) - considering evicting", serviceName, wantLb, *existingProbe.Name)
+			keepProbe := false
+			for _, expectedProbe := range expectedProbes {
```
looks to me like you can extract a `findProbe` method and use it here and below (if not other places)

brendandburns commented on pkg/cloudprovider/providers/azure/azure_loadbalancer.go:
```diff
@@ -0,0 +1,610 @@
+		updatedRules = *lb.Properties.LoadBalancingRules
+	}
+	// update rules: remove unwanted
+	for i := len(updatedRules) - 1; i >= 0; i-- {
```
same here for extracting a method.

brendandburns commented on pkg/cloudprovider/providers/azure/azure_loadbalancer.go:
```diff
@@ -0,0 +1,610 @@
+			glog.V(10).Infof("reconcile(%s)(%t): sg rule(%s) - considering evicting", serviceName, wantLb, *existingRule.Name)
+			keepRule := false
+
+			for _, expectedRule := range expectedSecurityRules {
```
here too.

brendandburns commented:
Basically LGTM, a few small re-factors to reduce some duplicate code.  Thanks for your patience.

--brendan

brendandburns commented:
(oh, and it needs a rebase)

colemickens commented:
Thanks, any thoughts on the two questions from my comment above (https://github.com/kubernetes/kubernetes/pull/28821#issuecomment-232933850)?

Specifically the logic for `Instances` `List` and whether or not you think it's worth building a switch to support cert auth for the Service Principal.

brendandburns commented:
@colemickens 

I think simply listing the resource group is fine for now.

Let's add CertAuth in a follow on PR, this one is large enough as it is.

colemickens commented:
* Factored out the `find{Probe,Rule,SecurityRule}` functions as suggested.
* Removed a couple other noisy log lines.
* Fixed the other small unnecessary variable.
* Implemented `Instances` `List`

Looks like the build is green as well.

rootfs commented:
@colemickens @brendandburns when do we expect this can be merged? thanks.

brendandburns commented:
LGTM.  Thanks for the patience.

mikedanese commented:
Can you squash fix-up commits @colemickens?

colemickens commented:
@brendandburns @mikedanese Rebase, squashed appropriately. Build looks green, but the LGTM label got dropped.

edevil commented:
Great work!

mikedanese commented:
Thanks @colemickens, this looks great

thockin commented:
This has plumbed clusterName throughout, but what is the definition of
cluster name?  Is it guaranteed to be set?  What format and length is it?
How does it get set?

On Tue, Jul 26, 2016 at 6:42 AM, Huamin Chen <notifications@github.com>
wrote:

> @colemickens <https://github.com/colemickens> @brendandburns
> <https://github.com/brendandburns> when do we expect this can be merged?
> thanks.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/28821#issuecomment-235270691>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVPx407fRzyld35ofWM3QBg8k4S7Uks5qZg6ugaJpZM4JKELW>
> .
>

colemickens commented:
@thockin: the `clusterName` is the user specified `--cluster-name` given to `kube-controller-manager`. It defaults to `kubernetes`. It is used for naming the load balancer in Azure so that two Kubernetes deployments could live side-by-side in a single Azure Resource Group. Since we don't create Azure Load Balancers per-service, and instead share one for the cluster, I needed some sort of name to be unique per cluster and chose `clusterName`. 

This `clusterName` was an existing parameter used in the `Routes` interface.

thockin commented:
huh, right you are.  I didn't know we had such a flag.  Thanks!

On Thu, Aug 18, 2016 at 9:06 PM, Cole Mickens <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin>: the clusterName is the user
> specified --cluster-name given to kube-controller-manager. It defaults to
> kubernetes. It is used for naming the load balancer in Azure so that two
> Kubernetes deployments could live side-by-side in a single Azure Resource
> Group. Since we don't create Azure Load Balancers per-service, and instead
> share one for the cluster, I needed some sort of name to be unique per
> cluster and chose clusterName.
>
> This clusterName was an existing parameter used in the Routes interface.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/28821#issuecomment-240922695>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVGXbyxQ7V9N8v7vODNGG8zGobSdyks5qhSvEgaJpZM4JKELW>
> .
>

jadbox commented:
How does this interact with the Azure VMSS cluster that may sit behind the load balancer? E.g. will I be able to tell the cluster to scale VMs up/down within kubernetes?

jadbox commented:
Never mind, I read the scope doc that says VMSS will not be considered: https://github.com/colemickens/azure-kubernetes-status
